File: Cloud Storage Samples/AzureBlob/java/azureblob.java
/*
 * Cloud Storage 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import cloudstorage.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class azureblob {

	public azureblob() {

		AzureBlob azureblob1 = new AzureBlob();
		String buffer = "";
		try {
			azureblob1.addAzureBlobEventListener(new AzureBlobEvents());
			System.out.print("Azure Account: ");
			azureblob1.setAccount(input());
			System.out.print("Access Key: ");
			azureblob1.setAccessKey(input());
			System.out.println("\nContainer List:\n");
			azureblob1.listContainers();
			printMenu();
			while (!buffer.equals("q")) {
				System.out.print("Enter command: ");
				buffer = input();
				if (buffer.toLowerCase().startsWith("cd")) {
					System.out.print("Container name: ");
					buffer = input();
					azureblob1.setContainer(buffer);
				} else if (buffer.toLowerCase().startsWith("ls")) {
					azureblob1.listBlobs();
				} else if (buffer.toLowerCase().startsWith("get")) {
					azureblob1.setLocalFile("");
					System.out.print("Blob name: ");
					azureblob1.getBlob(input());
					System.out.println("Contents of the file:\n"
							+ new String(azureblob1.getBlobData()));
				} else if (buffer.toLowerCase().startsWith("put")) {
					String rem_object;
					System.out.print("Name of new blob: ");
					rem_object = input();
					System.out.print("Local file to upload: ");
					azureblob1.setLocalFile(input());
					azureblob1.createBlob(rem_object, 0, -1);
					System.out.println("Blob created.");
				} else if (buffer.toLowerCase().equals("q")) {
					System.exit(0);
				} else if (buffer.toLowerCase().equals("?")) {
					printMenu();
				}
			}
		} catch (CloudStorageException e) {
			System.out.println(e.getMessage());
			System.exit(e.getCode());
			return;
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}

	}

	private String input() throws IOException {
		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		return bf.readLine();
	}

	private void printMenu() {
		System.out.println("\n?\t-\tHelp\n" + "cd\t-\tChange to container\n"
				+ "ls\t-\tList Blobs\n" + "get\t-\tGet remoteFile\n"
				+ "put\t-\tPut localFile\n" + "q\t-\tQuit\n");
	}

	public static void main(String[] args) {
		try {
			new azureblob();
		} catch (Exception ex) {
			System.out.println("Exception: " + ex.getMessage());
		}
	}

}

class AzureBlobEvents extends DefaultAzureBlobEventListener {
	public void blobList(AzureBlobBlobListEvent arg) {
		System.out.println(arg.name);
	}
	public void containerList(AzureBlobContainerListEvent arg) {
		System.out.println(arg.name);
	}
	public void SSLServerAuthentication(AzureBlobSSLServerAuthenticationEvent arg) {
		arg.accept = true;
	}
}




class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}

File: Cloud Storage Samples/AzureBlob/c++/azureblob.cpp
/*
 * Cloud Storage 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#define LINE_LEN 100



class MyBlob : public AzureBlob
{
public:

	virtual int FireContainerList(AzureBlobContainerListEventParams *e)
	{
		printf("%s\n", e->Name);
		return 0;
	}

	virtual int FireBlobList(AzureBlobBlobListEventParams *e)
	{
		printf("%s\n", e->Name);
		return 0;
	}

	virtual int FireError(AzureBlobErrorEventParams *e)
	{
		printf("Error Code: %d, Description: %s", e->ErrorCode, e->Description);
		return 0;
	}
};

void printoptions()
{
	printf( "Type ? to select this menu or a number 1-5 to perform one of the following actions:\n"
	        "1.  List Blobs\n"
	        "2.  Get Blob\n"
	        "3.  Put Blob\n"
	        "4.  Quit\n");
}


int main(int argc, char **argv)
{
	MyBlob blob;
	char buffer[LINE_LEN+1];
	int ret_code = 0;
	int size = 0;

	printf("Azure Account: ");
	scanf("%80s",buffer);
	blob.SetAccount(buffer);

	printf("Access Key: ");
	scanf("%100s",buffer);
	blob.SetAccessKey(buffer);

	printf("Listing Available Containers ...\n\n");
	if( ret_code = blob.ListContainers() ) goto error;

	printf("\nBlob Container to Examine: ");
	scanf("%80s",buffer);
	blob.SetContainer(buffer);

	printoptions();
	while (1)
	{
		printf( "blob> " );
		scanf("%80s",buffer);

		if ( ! strcmp(buffer, "?") )
		{
			printoptions();
		}

		else if ( ! strcmp(buffer, "1") )
		{
			if( ret_code = blob.ListBlobs() ) goto error;
		}

		else if ( ! strcmp(buffer, "2") )
		{
			printf("Which Blob?: ");
			scanf("%80s",buffer);
			if ( ret_code = blob.GetBlob(buffer) ) goto error;
			char * data;
			if ( ret_code = blob.GetBlobData(data, size) ) goto error;
			printf("Contents of blob %s :\n",buffer);
			for(int i=0; i < size; i++)
			{
				printf("%c",data[i]);
			}
			printf("\n");
		}

		else if ( ! strcmp(buffer, "3") )
		{
			char blobname[81];
			printf("Name of new blob: ");
			scanf("%80s",blobname);
			printf("Local file to upload: ");
			scanf("%80s",buffer);
			blob.SetLocalFile(buffer);
			if( ret_code = blob.CreateBlob(blobname, 0, -1) ) goto error;
			printf("Blob %s created.\n",blobname);
			blob.SetLocalFile("");
		}

		else if ( ! strcmp(buffer, "4") )
		{
			exit(0);
		}

		else if ( ! strcmp(buffer, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Please select a number from [1-4].\n" );
		} // end of command checking

error:
		if(ret_code)
		{
			printf("\nError: %d", ret_code);
			if (blob.GetLastError())
			{
				printf( " \"%s\"\n", blob.GetLastError() );
			}
		}

	}  // end of main while loop
	return ret_code;
}



